File: android/src/main/java/ai/telenow/sdk/Dsp.kt
// ai.telenow.sdk — audio DSP (Kotlin port of telenow-audio-core / @telenow/client).
// Faithful to the tested Swift/TS/Rust references.
package ai.telenow.sdk

import kotlin.math.abs
import kotlin.math.log10
import kotlin.math.roundToInt
import kotlin.math.sqrt

object Dsp {
    fun le16ToShorts(bytes: ByteArray): ShortArray {
        val n = bytes.size / 2
        val out = ShortArray(n)
        for (i in 0 until n) {
            out[i] = ((bytes[i * 2].toInt() and 0xff) or ((bytes[i * 2 + 1].toInt() and 0xff) shl 8)).toShort()
        }
        return out
    }

    fun shortsToLe16(samples: ShortArray): ByteArray {
        val out = ByteArray(samples.size * 2)
        for (i in samples.indices) {
            val s = samples[i].toInt()
            out[i * 2] = (s and 0xff).toByte()
            out[i * 2 + 1] = ((s shr 8) and 0xff).toByte()
        }
        return out
    }

    fun mulawByteToPcm16(byte: Int): Short {
        val u = byte.inv() and 0xff
        val sign = u and 0x80
        val exponent = (u shr 4) and 0x07
        val mantissa = u and 0x0f
        var sample = ((mantissa shl 3) + 0x84) shl exponent
        sample -= 0x84
        return (if (sign != 0) -sample else sample).toShort()
    }

    fun linear16ToMulawByte(sample: Short): Byte {
        val bias = 0x84
        val clip = 32635
        var s = sample.toInt()
        var sign = 0
        if (s < 0) { s = -s; sign = 0x80 }
        if (s > clip) s = clip
        s += bias
        var exponent = 7
        var mask = 0x4000
        while ((s and mask) == 0 && exponent > 0) { exponent--; mask = mask shr 1 }
        val mantissa = (s shr (exponent + 3)) and 0x0f
        return (((sign or (exponent shl 4) or mantissa).inv()) and 0xff).toByte()
    }

    fun rmsDbfs(samples: ShortArray): Double {
        if (samples.isEmpty()) return Double.NEGATIVE_INFINITY
        var sum = 0.0
        for (s in samples) { val v = s / 32768.0; sum += v * v }
        val rms = sqrt(sum / samples.size)
        return if (rms > 0) 20 * log10(rms) else Double.NEGATIVE_INFINITY
    }

    fun resample(input: ShortArray, fromHz: Int, toHz: Int): ShortArray {
        if (fromHz == toHz || input.isEmpty()) return input
        val outLen = maxOf(1, input.size * toHz / fromHz)
        val step = fromHz.toDouble() / toHz
        val out = ShortArray(outLen)
        for (i in 0 until outLen) {
            val pos = i * step
            val i0 = pos.toInt()
            val i1 = minOf(i0 + 1, input.size - 1)
            val frac = pos - i0
            out[i] = (input[i0] * (1 - frac) + input[i1] * frac).roundToInt().toShort()
        }
        return out
    }
}

enum class JitterAction { PLAY, UNDERRUN, OVERRUN }

data class JitterDecision(val startAt: Double, val action: JitterAction, val bufferedSec: Double, val targetSec: Double)

data class JitterOptions(
    val minTargetSec: Double = 0.06,
    val maxTargetSec: Double = 0.4,
    val initialTargetSec: Double = 0.12,
    val jitterGain: Double = 3.0,
)

/** Adaptive jitter buffer — port of the tested reference. */
class AdaptiveJitterBuffer(private val o: JitterOptions = JitterOptions()) {
    private var target = o.initialTargetSec
    private var playCursor: Double? = null
    private var prevArrival: Double? = null
    private var jitterEst = 0.0

    fun reset() { playCursor = null; prevArrival = null; jitterEst = 0.0 }
    fun buffered(now: Double): Double = playCursor?.let { maxOf(0.0, it - now) } ?: 0.0
    fun targetDepth(): Double = target

    fun schedule(now: Double, frameDur: Double, arrival: Double): JitterDecision {
        prevArrival?.let { prev ->
            val inter = arrival - prev
            val dev = abs(inter - frameDur)
            jitterEst += (dev - jitterEst) / 16
            val desired = minOf(o.maxTargetSec, maxOf(o.minTargetSec, o.minTargetSec + o.jitterGain * jitterEst))
            target += (desired - target) * 0.1
        }
        prevArrival = arrival

        val action: JitterAction
        val startAt: Double
        val pc = playCursor
        when {
            pc == null -> { action = JitterAction.PLAY; startAt = now + target }
            pc < now -> { action = JitterAction.UNDERRUN; startAt = now + target }
            pc - now > o.maxTargetSec -> { action = JitterAction.OVERRUN; startAt = now + target }
            else -> { action = JitterAction.PLAY; startAt = pc }
        }
        playCursor = startAt + frameDur
        return JitterDecision(startAt, action, startAt + frameDur - now, target)
    }
}

File: flutter/lib/src/dsp.dart
// Telenow Voice SDK — audio DSP (Dart port of telenow-audio-core).
import 'dart:math' as math;
import 'dart:typed_data';

class Dsp {
  static Int16List le16ToShorts(Uint8List bytes) {
    final n = bytes.length ~/ 2;
    final out = Int16List(n);
    final bd = ByteData.sublistView(bytes);
    for (var i = 0; i < n; i++) {
      out[i] = bd.getInt16(i * 2, Endian.little);
    }
    return out;
  }

  static Uint8List shortsToLe16(Int16List samples) {
    final out = Uint8List(samples.length * 2);
    final bd = ByteData.sublistView(out);
    for (var i = 0; i < samples.length; i++) {
      bd.setInt16(i * 2, samples[i], Endian.little);
    }
    return out;
  }

  static int mulawByteToPcm16(int byte) {
    final u = (~byte) & 0xff;
    final sign = u & 0x80;
    final exponent = (u >> 4) & 0x07;
    final mantissa = u & 0x0f;
    var sample = ((mantissa << 3) + 0x84) << exponent;
    sample -= 0x84;
    return sign != 0 ? -sample : sample;
  }

  static int linear16ToMulawByte(int sample) {
    const bias = 0x84;
    const clip = 32635;
    var s = sample;
    var sign = 0;
    if (s < 0) {
      s = -s;
      sign = 0x80;
    }
    if (s > clip) s = clip;
    s += bias;
    var exponent = 7;
    var mask = 0x4000;
    while ((s & mask) == 0 && exponent > 0) {
      exponent--;
      mask >>= 1;
    }
    final mantissa = (s >> (exponent + 3)) & 0x0f;
    return (~(sign | (exponent << 4) | mantissa)) & 0xff;
  }

  static double rmsDbfs(Int16List samples) {
    if (samples.isEmpty) return double.negativeInfinity;
    var sum = 0.0;
    for (final s in samples) {
      final v = s / 32768.0;
      sum += v * v;
    }
    final rms = math.sqrt(sum / samples.length);
    return rms > 0 ? 20 * (math.log(rms) / math.ln10) : double.negativeInfinity;
  }

  static Int16List resample(Int16List input, int fromHz, int toHz) {
    if (fromHz == toHz || input.isEmpty) return input;
    final outLen = math.max(1, input.length * toHz ~/ fromHz);
    final step = fromHz / toHz;
    final out = Int16List(outLen);
    for (var i = 0; i < outLen; i++) {
      final pos = i * step;
      final i0 = pos.floor();
      final i1 = math.min(i0 + 1, input.length - 1);
      final frac = pos - i0;
      out[i] = (input[i0] * (1 - frac) + input[i1] * frac).round();
    }
    return out;
  }
}

enum JitterAction { play, underrun, overrun }

class JitterDecision {
  final double startAt;
  final JitterAction action;
  final double bufferedSec;
  final double targetSec;
  JitterDecision(this.startAt, this.action, this.bufferedSec, this.targetSec);
}

class JitterOptions {
  final double minTargetSec;
  final double maxTargetSec;
  final double initialTargetSec;
  final double jitterGain;
  const JitterOptions({
    this.minTargetSec = 0.06,
    this.maxTargetSec = 0.4,
    this.initialTargetSec = 0.12,
    this.jitterGain = 3.0,
  });
}

/// Adaptive jitter buffer — port of the tested reference.
class AdaptiveJitterBuffer {
  final JitterOptions o;
  double _target;
  double? _playCursor;
  double? _prevArrival;
  double _jitterEst = 0;

  AdaptiveJitterBuffer([JitterOptions? opts])
      : o = opts ?? const JitterOptions(),
        _target = (opts ?? const JitterOptions()).initialTargetSec;

  void reset() {
    _playCursor = null;
    _prevArrival = null;
    _jitterEst = 0;
  }

  double buffered(double now) => _playCursor == null ? 0 : math.max(0, _playCursor! - now);
  double targetDepth() => _target;

  JitterDecision schedule(double now, double frameDur, double arrival) {
    final prev = _prevArrival;
    if (prev != null) {
      final inter = arrival - prev;
      final dev = (inter - frameDur).abs();
      _jitterEst += (dev - _jitterEst) / 16;
      final desired = math.min(
        o.maxTargetSec,
        math.max(o.minTargetSec, o.minTargetSec + o.jitterGain * _jitterEst),
      );
      _target += (desired - _target) * 0.1;
    }
    _prevArrival = arrival;

    final JitterAction action;
    final double startAt;
    final pc = _playCursor;
    if (pc == null) {
      action = JitterAction.play;
      startAt = now + _target;
    } else if (pc < now) {
      action = JitterAction.underrun;
      startAt = now + _target;
    } else if (pc - now > o.maxTargetSec) {
      action = JitterAction.overrun;
      startAt = now + _target;
    } else {
      action = JitterAction.play;
      startAt = pc;
    }
    _playCursor = startAt + frameDur;
    return JitterDecision(startAt, action, startAt + frameDur - now, _target);
  }
}
